File: agents/windows/src/N3NAgent.Core/Protocol/GatewayProtocol.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace N3NAgent.Core.Protocol;

/// <summary>
/// Gateway message types
/// </summary>
public enum GatewayMessageType
{
    [JsonPropertyName("req")] Request,
    [JsonPropertyName("res")] Response,
    [JsonPropertyName("event")] Event
}

/// <summary>
/// Base gateway message
/// </summary>
public class GatewayMessage
{
    [JsonPropertyName("type")]
    public string Type { get; set; } = "";

    [JsonPropertyName("id")]
    public string? Id { get; set; }
}

/// <summary>
/// Gateway request message
/// </summary>
public class GatewayRequest : GatewayMessage
{
    [JsonPropertyName("method")]
    public string Method { get; set; } = "";

    [JsonPropertyName("params")]
    public Dictionary<string, object>? Params { get; set; }

    public GatewayRequest()
    {
        Type = "req";
        Id = Guid.NewGuid().ToString();
    }

    public static GatewayRequest Create(string method, Dictionary<string, object>? parameters = null)
    {
        return new GatewayRequest
        {
            Method = method,
            Params = parameters
        };
    }

    public static GatewayRequest HandshakeAuth(string deviceId, string deviceToken, List<string> capabilities)
    {
        return Create("handshake.auth", new Dictionary<string, object>
        {
            ["deviceId"] = deviceId,
            ["deviceToken"] = deviceToken,
            ["capabilities"] = capabilities,
            ["client"] = new Dictionary<string, object>
            {
                ["version"] = "1.0.0",
                ["platform"] = "windows",
                ["arch"] = Environment.Is64BitOperatingSystem ? "x64" : "x86"
            }
        });
    }

    public static GatewayRequest NodeRegister(List<string> capabilities)
    {
        return Create("node.register", new Dictionary<string, object>
        {
            ["capabilities"] = capabilities
        });
    }

    public static GatewayRequest Ping()
    {
        return Create("ping");
    }

    public static GatewayRequest InvokeResult(string invokeId, object? result, string? error)
    {
        var parameters = new Dictionary<string, object>
        {
            ["invokeId"] = invokeId
        };

        if (result != null)
            parameters["result"] = result;

        if (error != null)
            parameters["error"] = error;

        return Create("node.invoke.result", parameters);
    }
}

/// <summary>
/// Gateway response message
/// </summary>
public class GatewayResponse : GatewayMessage
{
    [JsonPropertyName("success")]
    public bool Success { get; set; }

    [JsonPropertyName("result")]
    public Dictionary<string, object>? Result { get; set; }

    [JsonPropertyName("error")]
    public ResponseError? Error { get; set; }

    public GatewayResponse()
    {
        Type = "res";
    }

    public class ResponseError
    {
        [JsonPropertyName("code")]
        public string Code { get; set; } = "";

        [JsonPropertyName("message")]
        public string Message { get; set; } = "";
    }

    public static GatewayResponse CreateSuccess(string id, Dictionary<string, object> result)
    {
        return new GatewayResponse
        {
            Id = id,
            Success = true,
            Result = result
        };
    }

    public static GatewayResponse CreateError(string id, string code, string message)
    {
        return new GatewayResponse
        {
            Id = id,
            Success = false,
            Error = new ResponseError { Code = code, Message = message }
        };
    }
}

/// <summary>
/// Gateway event message
/// </summary>
public class GatewayEvent : GatewayMessage
{
    [JsonPropertyName("event")]
    public string Event { get; set; } = "";

    [JsonPropertyName("payload")]
    public Dictionary<string, object>? Payload { get; set; }

    public GatewayEvent()
    {
        Type = "event";
    }

    public static GatewayEvent Create(string eventName, Dictionary<string, object>? payload = null)
    {
        return new GatewayEvent
        {
            Event = eventName,
            Payload = payload
        };
    }

    // Common events
    public const string HandshakeChallenge = "handshake.challenge";
    public const string Connected = "connected";
    public const string Disconnected = "disconnected";
    public const string NodeInvoke = "node.invoke";
    public const string Ping = "ping";
    public const string Pong = "pong";
}

/// <summary>
/// Node invoke request (from server)
/// </summary>
public class NodeInvokeRequest
{
    [JsonPropertyName("invokeId")]
    public string InvokeId { get; set; } = "";

    [JsonPropertyName("capability")]
    public string Capability { get; set; } = "";

    [JsonPropertyName("args")]
    public Dictionary<string, object>? Args { get; set; }
}

File: agents/macos/Sources/N3NAgentCore/Protocol/GatewayProtocol.swift
import Foundation

/// Base gateway message type
public enum GatewayMessageType: String, Codable {
    case request = "req"
    case response = "res"
    case event = "event"
}

/// Gateway message base
public struct GatewayMessage: Codable {
    public let type: String
    public let id: String?

    public init(type: GatewayMessageType, id: String? = nil) {
        self.type = type.rawValue
        self.id = id
    }
}

/// Gateway request message
public struct GatewayRequest: Codable {
    public let type: String
    public let id: String
    public let method: String
    public let params: [String: AnyCodable]?

    public init(method: String, params: [String: AnyCodable]? = nil) {
        self.type = GatewayMessageType.request.rawValue
        self.id = UUID().uuidString
        self.method = method
        self.params = params
    }

    public static func handshakeAuth(
        deviceId: String,
        deviceToken: String,
        capabilities: [String]
    ) -> GatewayRequest {
        return GatewayRequest(
            method: "handshake.auth",
            params: [
                "deviceId": AnyCodable(deviceId),
                "deviceToken": AnyCodable(deviceToken),
                "capabilities": AnyCodable(capabilities),
                "client": AnyCodable([
                    "version": "1.0.0",
                    "platform": "macos",
                    "arch": getArchitecture()
                ])
            ]
        )
    }

    public static func nodeRegister(capabilities: [String]) -> GatewayRequest {
        return GatewayRequest(
            method: "node.register",
            params: [
                "capabilities": AnyCodable(capabilities)
            ]
        )
    }

    public static func ping() -> GatewayRequest {
        return GatewayRequest(method: "ping")
    }

    public static func invokeResult(invokeId: String, result: Any?, error: String?) -> GatewayRequest {
        var params: [String: AnyCodable] = [
            "invokeId": AnyCodable(invokeId)
        ]

        if let result = result {
            params["result"] = AnyCodable(result)
        }

        if let error = error {
            params["error"] = AnyCodable(error)
        }

        return GatewayRequest(
            method: "node.invoke.result",
            params: params
        )
    }

    private static func getArchitecture() -> String {
        #if arch(arm64)
        return "arm64"
        #elseif arch(x86_64)
        return "x86_64"
        #else
        return "unknown"
        #endif
    }
}

/// Gateway response message
public struct GatewayResponse: Codable {
    public let type: String
    public let id: String
    public let success: Bool
    public let result: [String: AnyCodable]?
    public let error: ResponseError?

    public struct ResponseError: Codable {
        public let code: String
        public let message: String

        public init(code: String, message: String) {
            self.code = code
            self.message = message
        }
    }

    public init(id: String, success: Bool, result: [String: AnyCodable]? = nil, error: ResponseError? = nil) {
        self.type = GatewayMessageType.response.rawValue
        self.id = id
        self.success = success
        self.result = result
        self.error = error
    }

    public static func success(id: String, result: [String: AnyCodable]) -> GatewayResponse {
        return GatewayResponse(id: id, success: true, result: result)
    }

    public static func error(id: String, code: String, message: String) -> GatewayResponse {
        return GatewayResponse(id: id, success: false, error: ResponseError(code: code, message: message))
    }
}

/// Gateway event message
public struct GatewayEvent: Codable {
    public let type: String
    public let event: String
    public let payload: [String: AnyCodable]?

    public init(event: String, payload: [String: AnyCodable]? = nil) {
        self.type = GatewayMessageType.event.rawValue
        self.event = event
        self.payload = payload
    }

    // Common events
    public static let handshakeChallenge = "handshake.challenge"
    public static let connected = "connected"
    public static let disconnected = "disconnected"
    public static let nodeInvoke = "node.invoke"
    public static let ping = "ping"
    public static let pong = "pong"
}

/// Node invoke request (from server)
public struct NodeInvokeRequest: Codable {
    public let invokeId: String
    public let capability: String
    public let args: [String: AnyCodable]?

    public init(invokeId: String, capability: String, args: [String: AnyCodable]? = nil) {
        self.invokeId = invokeId
        self.capability = capability
        self.args = args
    }
}

// MARK: - AnyCodable Helper

/// A type-erased Codable value.
public struct AnyCodable: Codable {
    public let value: Any

    public init(_ value: Any) {
        self.value = value
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()

        if container.decodeNil() {
            self.value = NSNull()
        } else if let bool = try? container.decode(Bool.self) {
            self.value = bool
        } else if let int = try? container.decode(Int.self) {
            self.value = int
        } else if let double = try? container.decode(Double.self) {
            self.value = double
        } else if let string = try? container.decode(String.self) {
            self.value = string
        } else if let array = try? container.decode([AnyCodable].self) {
            self.value = array.map { $0.value }
        } else if let dictionary = try? container.decode([String: AnyCodable].self) {
            self.value = dictionary.mapValues { $0.value }
        } else {
            throw DecodingError.dataCorruptedError(in: container, debugDescription: "Cannot decode value")
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()

        switch value {
        case is NSNull:
            try container.encodeNil()
        case let bool as Bool:
            try container.encode(bool)
        case let int as Int:
            try container.encode(int)
        case let double as Double:
            try container.encode(double)
        case let string as String:
            try container.encode(string)
        case let array as [Any]:
            try container.encode(array.map { AnyCodable($0) })
        case let dictionary as [String: Any]:
            try container.encode(dictionary.mapValues { AnyCodable($0) })
        default:
            throw EncodingError.invalidValue(value, EncodingError.Context(
                codingPath: container.codingPath,
                debugDescription: "Cannot encode value"
            ))
        }
    }

    public func asString() -> String? {
        return value as? String
    }

    public func asInt() -> Int? {
        return value as? Int
    }

    public func asBool() -> Bool? {
        return value as? Bool
    }

    public func asDouble() -> Double? {
        return value as? Double
    }

    public func asArray() -> [Any]? {
        return value as? [Any]
    }

    public func asDictionary() -> [String: Any]? {
        return value as? [String: Any]
    }
}
